File: apps/ios/FishKit/Sources/ChatData/Adapters/SharedContentThumbnailStore.swift
import ChatCore
import CryptoKit
import Foundation

/// Opaque identity for one displayed thumbnail. The raw values never appear
/// in the filesystem path or in the value's textual representation.
public struct SharedContentThumbnailKey: Hashable, Sendable, CustomStringConvertible {
    public let ownerIdentityId: String
    public let conversationId: String
    public let itemId: String
    public let contentVersion: String
    public let identityGeneration: Int

    public let ownerFingerprint: String
    public let opaqueRelativePath: String

    public init(
        _ ownerIdentityId: String,
        _ conversationId: String,
        _ itemId: String,
        _ contentVersion: String,
        identityGeneration: Int
    ) {
        precondition(!ownerIdentityId.isEmpty)
        precondition(!conversationId.isEmpty)
        precondition(!itemId.isEmpty)
        precondition(!contentVersion.isEmpty)
        precondition(identityGeneration > 0)
        self.ownerIdentityId = ownerIdentityId
        self.conversationId = conversationId
        self.itemId = itemId
        self.contentVersion = contentVersion
        self.identityGeneration = identityGeneration
        ownerFingerprint = Self.sha256(ownerIdentityId)
        let conversationFingerprint = Self.sha256(conversationId)
        let itemFingerprint = Self.sha256(itemId)
        let versionFingerprint = Self.sha256("\(contentVersion)|\(identityGeneration)")
        opaqueRelativePath = "\(ownerFingerprint)/\(conversationFingerprint)/\(itemFingerprint)-\(versionFingerprint).thumb"
    }

    public var description: String {
        "SharedContentThumbnailKey(ownerFingerprint=\(ownerFingerprint), opaqueRelativePath=\(opaqueRelativePath))"
    }

    private static func sha256(_ value: String) -> String {
        SHA256.hash(data: Data(value.utf8)).map { String(format: "%02x", $0) }.joined()
    }
}

/// Thumbnail bytes are copied at the boundary and carry no URL/path metadata.
public struct SharedContentThumbnailBytes: Equatable, Sendable, CustomStringConvertible {
    public let data: Data
    public var size: Int { data.count }

    public init(_ data: Data) {
        self.data = Data(data)
    }

    public var description: String { "SharedContentThumbnailBytes(size=\(size))" }
}

/// Explicit marker for the only durable media intent.
public enum SharedContentDisplayedIntent: Sendable, Equatable {
    case displayedThumbnail
}

/// Owner-scoped displayed-thumbnail cache. Lookahead and selected-full bytes
/// remain in bounded actor memory; only `confirmDisplayed` writes a file.
public actor SharedContentThumbnailStore {
    public static let maximumThumbnailBytes: Int64 = 8 * 1024 * 1024
    public static let maximumStagedBytes: Int64 = 16 * 1024 * 1024
    public static let defaultMaximumBytes: Int64 = 64 * 1024 * 1024
    public static let defaultInactivity: TimeInterval = 30 * 24 * 60 * 60

    public let root: URL
    public let isExcludedFromBackup: Bool
    private let fileManager: FileManager
    private let now: @Sendable () -> Date
    private let canonicalRoot: URL
    private var staged: [SharedContentThumbnailKey: SharedContentThumbnailBytes] = [:]
    private var stagedByteCount: Int64 = 0
    private var revokedThroughGeneration: [String: Int] = [:]

    public init(
        root: URL? = nil,
        fileManager: FileManager = .default,
        now: @escaping @Sendable () -> Date = Date.init
    ) throws {
        self.fileManager = fileManager
        self.now = now
        if let root {
            self.root = root
        } else {
            guard let caches = fileManager.urls(for: .cachesDirectory, in: .userDomainMask).first else {
                throw SharedContentThumbnailStoreFailure.unavailable
            }
            self.root = caches.appending(path: "SharedContentThumbnails", directoryHint: .isDirectory)
        }
        try fileManager.createDirectory(at: self.root, withIntermediateDirectories: true)
        self.canonicalRoot = self.root.standardizedFileURL
        try Self.protectAndExclude(self.root)
        self.isExcludedFromBackup = true
    }

    @discardableResult
    public func stageLookahead(
        _ key: SharedContentThumbnailKey,
        bytes: Data
    ) -> Bool {
        stage(key, bytes: bytes, intent: .lookaheadThumbnail)
    }

    @discardableResult
    public func stage(
        _ key: SharedContentThumbnailKey,
        bytes: Data,
        intent: SharedContentFetchIntent
    ) -> Bool {
        guard key.identityGeneration > revokedThroughGeneration[key.ownerIdentityId, default: 0],
              intent != .selectedFullContent,
              !bytes.isEmpty,
              Int64(bytes.count) <= Self.maximumThumbnailBytes
        else { return false }

        if let old = staged.removeValue(forKey: key) { stagedByteCount -= Int64(old.size) }
        staged[key] = SharedContentThumbnailBytes(bytes)
        stagedByteCount += Int64(bytes.count)
        while stagedByteCount > Self.maximumStagedBytes, let first = staged.first {
            staged.removeValue(forKey: first.key)
            stagedByteCount -= Int64(first.value.size)
        }
        return true
    }

    @discardableResult
    public func confirmDisplayed(_ key: SharedContentThumbnailKey) -> Bool {
        guard key.identityGeneration > revokedThroughGeneration[key.ownerIdentityId, default: 0] else {
            removeStaged(key)
            return false
        }
        guard let bytes = staged[key], writeDisplayedAtomically(key, bytes: bytes) else { return false }
        removeStaged(key)
        return true
    }

    public func readDisplayed(_ key: SharedContentThumbnailKey) -> SharedContentThumbnailBytes? {
        guard key.identityGeneration > revokedThroughGeneration[key.ownerIdentityId, default: 0],
              let url = containedFile(for: key),
              let values = try? url.resourceValues(forKeys: [.isRegularFileKey, .fileSizeKey]),
              values.isRegularFile == true,
              let fileSize = values.fileSize,
              fileSize > 0,
              Int64(fileSize) <= Self.maximumThumbnailBytes,
              let data = try? Data(contentsOf: url, options: .mappedIfSafe)
        else { return nil }
        try? fileManager.setAttributes([.modificationDate: now()], ofItemAtPath: url.path)
        return SharedContentThumbnailBytes(data)
    }

    /// Reads staged or durable bytes without promoting lookahead work.
    public func readRenderable(_ key: SharedContentThumbnailKey) -> SharedContentThumbnailBytes? {
        guard key.identityGeneration > revokedThroughGeneration[key.ownerIdentityId, default: 0]
        else { return nil }
        if let value = staged[key] {
            return SharedContentThumbnailBytes(value.data)
        }
        return readDisplayed(key)
    }

    public func prune(
        ownerIdentityId: String,
        maxBytes: Int64 = SharedContentThumbnailStore.defaultMaximumBytes,
        inactiveAfter: TimeInterval = SharedContentThumbnailStore.defaultInactivity
    ) {
        guard !ownerIdentityId.isEmpty, maxBytes >= 0, inactiveAfter >= 0,
              let ownerRoot = containedOwnerRoot(ownerIdentityId)
        else { return }
        let cutoff = now().addingTimeInterval(-inactiveAfter)
        let files = thumbnailFiles(under: ownerRoot)
        files.filter { modificationDate(for: $0) <= cutoff }.forEach { try? fileManager.removeItem(at: $0) }

        var remaining = thumbnailFiles(under: ownerRoot)
        var total = remaining.reduce(Int64(0)) { $0 + Int64((try? $1.resourceValues(forKeys: [.fileSizeKey]).fileSize) ?? 0) }
        for file in remaining.sorted(by: { modificationDate(for: $0) < modificationDate(for: $1) }) where total > maxBytes {
            let size = Int64((try? file.resourceValues(forKeys: [.fileSizeKey]).fileSize) ?? 0)
            try? fileManager.removeItem(at: file)
            total -= size
        }
        remaining = thumbnailFiles(under: ownerRoot)
        removeEmptyDirectories(under: ownerRoot)
        _ = remaining
    }

    @discardableResult
    public func purgeOwner(ownerIdentityId: String) -> Bool {
        guard !ownerIdentityId.isEmpty else { return false }
        staged = staged.filter { key, bytes in
            guard key.ownerIdentityId != ownerIdentityId else {
                stagedByteCount -= Int64(bytes.size)
                return false
            }
            return true
        }
        guard let ownerRoot = containedOwnerRoot(ownerIdentityId) else { return false }
        deleteContainedTree(ownerRoot)
        return !fileManager.fileExists(atPath: ownerRoot.path)
    }

    @discardableResult
    public func purgeConversation(ownerIdentityId: String, conversationId: String) -> Bool {
        guard !ownerIdentityId.isEmpty, !conversationId.isEmpty else { return false }
        staged = staged.filter { key, bytes in
            guard key.ownerIdentityId != ownerIdentityId || key.conversationId != conversationId else {
                stagedByteCount -= Int64(bytes.size)
                return false
            }
            return true
        }
        let conversationRoot = root
            .appending(path: SHA256.hash(data: Data(ownerIdentityId.utf8)).map { String(format: "%02x", $0) }.joined())
            .appending(path: SHA256.hash(data: Data(conversationId.utf8)).map { String(format: "%02x", $0) }.joined())
        guard isContained(conversationRoot) else { return false }
        deleteContainedTree(conversationRoot)
        return !fileManager.fileExists(atPath: conversationRoot.path)
    }

    public var stagedCount: Int { staged.count }

    public func stagedCount(ownerIdentityId: String) -> Int {
        staged.keys.lazy.filter { $0.ownerIdentityId == ownerIdentityId }.count
    }

    public func revokeIdentityGeneration(ownerIdentityId: String, through generation: Int) {
        guard !ownerIdentityId.isEmpty, generation > 0 else { return }
        revokedThroughGeneration[ownerIdentityId] = max(
            revokedThroughGeneration[ownerIdentityId, default: 0],
            generation
        )
        staged = staged.filter { key, bytes in
            guard key.ownerIdentityId != ownerIdentityId || key.identityGeneration > generation else {
                stagedByteCount -= Int64(bytes.size)
                return false
            }
            return true
        }
    }

    public func persistedFileCount(ownerIdentityId: String) -> Int {
        guard let ownerRoot = containedOwnerRoot(ownerIdentityId) else { return 0 }
        return thumbnailFiles(under: ownerRoot).count
    }

    public func persistedByteCount(ownerIdentityId: String) -> Int64 {
        guard let ownerRoot = containedOwnerRoot(ownerIdentityId) else { return 0 }
        return thumbnailFiles(under: ownerRoot).reduce(Int64(0)) {
            $0 + Int64((try? $1.resourceValues(forKeys: [.fileSizeKey]).fileSize) ?? 0)
        }
    }

    private func writeDisplayedAtomically(
        _ key: SharedContentThumbnailKey,
        bytes: SharedContentThumbnailBytes
    ) -> Bool {
        guard let destination = containedFile(for: key),
              bytes.size > 0,
              Int64(bytes.size) <= Self.maximumThumbnailBytes
        else { return false }
        do {
            try fileManager.createDirectory(at: destination.deletingLastPathComponent(), withIntermediateDirectories: true)
            try Self.protectAndExclude(destination.deletingLastPathComponent())
            let temporary = destination.deletingLastPathComponent()
                .appending(path: ".\(destination.lastPathComponent).tmp")
            try bytes.data.write(to: temporary, options: [.atomic, .completeFileProtectionUntilFirstUserAuthentication])
            try Self.protectAndExclude(temporary)
            if fileManager.fileExists(atPath: destination.path) { try fileManager.removeItem(at: destination) }
            try fileManager.moveItem(at: temporary, to: destination)
            try Self.protectAndExclude(destination)
            return true
        } catch {
            let temporary = destination.deletingLastPathComponent()
                .appending(path: ".\(destination.lastPathComponent).tmp")
            try? fileManager.removeItem(at: temporary)
            return false
        }
    }

    private func removeStaged(_ key: SharedContentThumbnailKey) {
        guard let bytes = staged.removeValue(forKey: key) else { return }
        stagedByteCount -= Int64(bytes.size)
    }

    private func containedFile(for key: SharedContentThumbnailKey) -> URL? {
        let candidate = canonicalRoot.appending(path: key.opaqueRelativePath)
        return isContained(candidate) ? candidate : nil
    }

    private func containedOwnerRoot(_ ownerIdentityId: String) -> URL? {
        let fingerprint = SHA256.hash(data: Data(ownerIdentityId.utf8)).map { String(format: "%02x", $0) }.joined()
        let candidate = canonicalRoot.appending(path: fingerprint)
        return isContained(candidate) ? candidate : nil
    }

    private func isContained(_ url: URL) -> Bool {
        let rootPath = canonicalRoot.resolvingSymlinksInPath().path
        let path = url.standardizedFileURL.resolvingSymlinksInPath().path
        return path == rootPath || path.hasPrefix(rootPath + "/")
    }

    private func thumbnailFiles(under directory: URL) -> [URL] {
        guard isContained(directory), let enumerator = fileManager.enumerator(
            at: directory,
            includingPropertiesForKeys: [.isRegularFileKey, .fileSizeKey, .contentModificationDateKey],
            options: [.skipsHiddenFiles]
        ) else { return [] }
        return enumerator.compactMap { value in
            guard let url = value as? URL, url.pathExtension == "thumb", isContained(url),
                  (try? url.resourceValues(forKeys: [.isRegularFileKey]).isRegularFile) == true
            else { return nil }
            return url
        }
    }

    private func modificationDate(for url: URL) -> Date {
        (try? url.resourceValues(forKeys: [.contentModificationDateKey]).contentModificationDate) ?? .distantPast
    }

    private func deleteContainedTree(_ directory: URL) {
        guard isContained(directory), fileManager.fileExists(atPath: directory.path),
              let enumerator = fileManager.enumerator(at: directory, includingPropertiesForKeys: nil)
        else { return }
        let entries = enumerator.compactMap { $0 as? URL }.sorted { $0.path.count > $1.path.count }
        for entry in entries where isContained(entry) { try? fileManager.removeItem(at: entry) }
        try? fileManager.removeItem(at: directory)
    }

    private func removeEmptyDirectories(under directory: URL) {
        guard isContained(directory), let enumerator = fileManager.enumerator(at: directory, includingPropertiesForKeys: [.isDirectoryKey]) else { return }
        for entry in enumerator.compactMap({ $0 as? URL }).sorted(by: { $0.path.count > $1.path.count }) where isContained(entry) {
            if (try? entry.resourceValues(forKeys: [.isDirectoryKey]).isDirectory) == true,
               (try? fileManager.contentsOfDirectory(at: entry, includingPropertiesForKeys: nil)).map(\.isEmpty) == true {
                try? fileManager.removeItem(at: entry)
            }
        }
    }

    private static func protectAndExclude(_ url: URL) throws {
        var values = URLResourceValues()
        values.isExcludedFromBackup = true
        var mutableUrl = url
        try mutableUrl.setResourceValues(values)
    }
}

public enum SharedContentThumbnailStoreFailure: Error, Equatable, Sendable {
    case unavailable
}

File: apps/android/data/chat/src/main/kotlin/space/fishhub/android/data/chat/sharedcontent/SharedContentThumbnailStore.kt
package space.fishhub.android.data.chat.sharedcontent

import android.content.Context
import java.io.File
import java.io.FileOutputStream
import java.nio.file.AtomicMoveNotSupportedException
import java.nio.file.Files
import java.nio.file.StandardCopyOption
import java.security.MessageDigest
import java.time.Duration
import java.time.Instant
import java.util.LinkedHashMap

/** Opaque identity for one displayed thumbnail on the device. */
data class SharedContentThumbnailKey(
    val ownerIdentityId: String,
    val conversationId: String,
    val itemId: String,
    val contentVersion: String,
) {
    init {
        require(ownerIdentityId.isNotBlank())
        require(conversationId.isNotBlank())
        require(itemId.isNotBlank())
        require(contentVersion.isNotBlank())
    }

    val ownerFingerprint: String = sha256(ownerIdentityId)
    private val conversationFingerprint: String = sha256(conversationId)
    private val itemFingerprint: String = sha256(itemId)
    private val contentVersionFingerprint: String = sha256(contentVersion)

    /** Contains no raw owner, conversation, item, or provider value. */
    val opaqueRelativePath: String
        get() = "$ownerFingerprint/$conversationFingerprint/$itemFingerprint-$contentVersionFingerprint.thumb"

    override fun toString(): String =
        "SharedContentThumbnailKey(ownerFingerprint=$ownerFingerprint, opaqueRelativePath=$opaqueRelativePath)"

    private companion object {
        fun sha256(value: String): String = MessageDigest.getInstance("SHA-256")
            .digest(value.toByteArray(Charsets.UTF_8))
            .joinToString("") { byte -> "%02x".format(byte.toInt() and 0xff) }
    }
}

/** A bounded in-memory or displayed-thumbnail byte value. */
class SharedContentThumbnailBytes(bytes: ByteArray) {
    val bytes: ByteArray = bytes.copyOf()
    val size: Int get() = bytes.size

    override fun equals(other: Any?): Boolean =
        other is SharedContentThumbnailBytes && bytes.contentEquals(other.bytes)

    override fun hashCode(): Int = bytes.contentHashCode()

    override fun toString(): String = "SharedContentThumbnailBytes(size=$size)"
}

/** Binary-compatibility marker for the original Wave 0 cache-key name. */
@Deprecated("Use SharedContentThumbnailKey.")
class ThumbnailCacheKey private constructor()

/**
 * Bounded, owner-scoped thumbnail context.
 *
 * Lookahead bytes are held only in [stagedLookahead]. The only path that writes
 * a thumbnail file is [confirmDisplayed], after the caller has confirmed that
 * the item was actually rendered. Full-content bytes have no API in this type.
 */
class SharedContentThumbnailStore internal constructor(
    private val root: File,
    private val now: () -> Instant,
) {
    constructor(
        context: Context,
        now: () -> Instant = Instant::now,
    ) : this(
        root = File(context.applicationContext.noBackupFilesDir, ROOT_DIRECTORY),
        now = now,
    )

    private val canonicalRoot = root.canonicalFile
    private val stateLock = Any()
    private val fileSystemLock = Any()
    private val stagedLookahead = LinkedHashMap<SharedContentThumbnailKey, SharedContentThumbnailBytes>()
    private var stagedBytes = 0L
    private var revisionSequence = 0L
    private val keyRevisions = mutableMapOf<SharedContentThumbnailKey, Long>()
    private val ownerPurgeEpochs = mutableMapOf<String, Long>()
    private val conversationPurgeEpochs = mutableMapOf<Pair<String, String>, Long>()

    /** noBackupFilesDir is excluded from Android backup and device transfer. */
    val isBackupExcluded: Boolean = true

    init {
        canonicalRoot.mkdirs()
    }

    fun readDisplayed(key: SharedContentThumbnailKey): SharedContentThumbnailBytes? {
        return synchronized(fileSystemLock) {
            val file = containedFile(key) ?: return@synchronized null
            if (!file.isFile || file.length() !in 1..MAX_THUMBNAIL_BYTES) {
                return@synchronized null
            }
            val value = runCatching {
                SharedContentThumbnailBytes(file.readBytes())
            }.getOrNull() ?: return@synchronized null
            file.setLastModified(now().toEpochMilli())
            value
        }
    }

    /**
     * Returns bytes that are safe to render without promoting lookahead work
     * to durable storage. Promotion remains exclusive to [confirmDisplayed].
     */
    fun readRenderable(key: SharedContentThumbnailKey): SharedContentThumbnailBytes? =
        synchronized(stateLock) {
            stagedLookahead[key]?.let { SharedContentThumbnailBytes(it.bytes) }
        } ?: readDisplayed(key)

    fun stageLookahead(
        key: SharedContentThumbnailKey,
        bytes: ByteArray,
    ): Boolean {
        if (bytes.isEmpty() || bytes.size.toLong() > MAX_THUMBNAIL_BYTES) return false
        synchronized(stateLock) {
            stagedLookahead.remove(key)?.let { stagedBytes -= it.size }
            revisionSequence += 1
            keyRevisions[key] = revisionSequence
            stagedLookahead[key] = SharedContentThumbnailBytes(bytes)
            stagedBytes += bytes.size
            while (stagedBytes > MAX_STAGED_BYTES) {
                val oldest = stagedLookahead.entries.firstOrNull() ?: break
                stagedLookahead.remove(oldest.key)
                keyRevisions.remove(oldest.key)
                stagedBytes -= oldest.value.size
            }
            checkStagedAccountingLocked()
        }
        return true
    }

    fun stageLookahead(
        ownerIdentityId: String,
        conversationId: String,
        itemId: String,
        contentVersion: String,
        bytes: ByteArray,
    ): Boolean = stageLookahead(
        SharedContentThumbnailKey(ownerIdentityId, conversationId, itemId, contentVersion),
        bytes,
    )

    /** Promotes staged bytes only after display confirmation. */
    fun confirmDisplayed(key: SharedContentThumbnailKey): Boolean {
        val pending = synchronized(stateLock) {
            val staged = stagedLookahead[key] ?: return false
            PendingPromotion(
                bytes = SharedContentThumbnailBytes(staged.bytes),
                keyRevision = keyRevisions[key] ?: return false,
                ownerPurgeEpoch = ownerPurgeEpochs[key.ownerIdentityId] ?: 0,
                conversationPurgeEpoch =
                    conversationPurgeEpochs[key.ownerIdentityId to key.conversationId] ?: 0,
            )
        }
        val written = synchronized(fileSystemLock) {
            writeDisplayedAtomically(key, pending.bytes)
        }
        if (!written) return false

        val accepted = synchronized(stateLock) {
            val current = stagedLookahead[key]
            val currentRevision = keyRevisions[key]
            val generationStillCurrent =
                (ownerPurgeEpochs[key.ownerIdentityId] ?: 0) == pending.ownerPurgeEpoch &&
                    (conversationPurgeEpochs[key.ownerIdentityId to key.conversationId] ?: 0) ==
                    pending.conversationPurgeEpoch
            if (generationStillCurrent &&
                currentRevision == pending.keyRevision &&
                current != null
            ) {
                stagedLookahead.remove(key)
                keyRevisions.remove(key)
                stagedBytes -= current.size
                checkStagedAccountingLocked()
                true
            } else {
                false
            }
        }
        if (!accepted) {
            synchronized(fileSystemLock) {
                containedFile(key)?.delete()
            }
        }
        return accepted
    }

    fun confirmDisplayed(
        ownerIdentityId: String,
        conversationId: String,
        itemId: String,
        contentVersion: String,
    ): Boolean = confirmDisplayed(
        SharedContentThumbnailKey(ownerIdentityId, conversationId, itemId, contentVersion),
    )

    /** Removes old thumbnails, then least-recently displayed files to meet the cap. */
    fun prune(
        ownerIdentityId: String,
        maxBytes: Long = DEFAULT_MAX_BYTES,
        inactiveAfter: Duration = DEFAULT_INACTIVITY,
    ) {
        require(ownerIdentityId.isNotBlank())
        require(maxBytes >= 0)
        require(!inactiveAfter.isNegative)
        synchronized(fileSystemLock) {
            val ownerRoot = ownerRoot(ownerIdentityId) ?: return
            if (!ownerRoot.isDirectory) return
            val cutoff = now().minus(inactiveAfter).toEpochMilli()
            val files = thumbnailFiles(ownerRoot)
            files.filter { it.lastModified() <= cutoff }.forEach(File::delete)

            var remaining = thumbnailFiles(ownerRoot)
            var total = remaining.sumOf(File::length)
            if (total > maxBytes) {
                remaining = remaining.sortedBy(File::lastModified)
                for (file in remaining) {
                    if (total <= maxBytes) break
                    total -= file.length()
                    file.delete()
                }
            }
            removeEmptyDirectories(ownerRoot)
        }
    }

    /** Removes all runtime and displayed bytes for a verified owner and proves absence. */
    fun purgeOwner(ownerIdentityId: String): Boolean {
        require(ownerIdentityId.isNotBlank())
        synchronized(stateLock) {
            ownerPurgeEpochs[ownerIdentityId] = (ownerPurgeEpochs[ownerIdentityId] ?: 0) + 1
            removeStagedLocked { key -> key.ownerIdentityId == ownerIdentityId }
        }
        return synchronized(fileSystemLock) {
            val ownerRoot = ownerRoot(ownerIdentityId) ?: return@synchronized false
            deleteContainedTree(ownerRoot)
            !ownerRoot.exists()
        }
    }

    fun purgeConversation(ownerIdentityId: String, conversationId: String): Boolean {
        require(ownerIdentityId.isNotBlank())
        require(conversationId.isNotBlank())
        synchronized(stateLock) {
            val namespace = ownerIdentityId to conversationId
            conversationPurgeEpochs[namespace] = (conversationPurgeEpochs[namespace] ?: 0) + 1
            removeStagedLocked { key ->
                key.ownerIdentityId == ownerIdentityId && key.conversationId == conversationId
            }
        }
        return synchronized(fileSystemLock) {
            val path = File(ownerRoot(ownerIdentityId), sha256(conversationId))
            if (!isContained(path)) return@synchronized false
            deleteContainedTree(path)
            !path.exists()
        }
    }

    fun stagedCount(): Int = synchronized(stateLock) { stagedLookahead.size }

    internal fun stagedByteCount(): Long = synchronized(stateLock) {
        checkStagedAccountingLocked()
        stagedBytes
    }

    fun persistedFileCount(ownerIdentityId: String): Int = synchronized(fileSystemLock) {
        ownerRoot(ownerIdentityId)?.let(::thumbnailFiles)?.size ?: 0
    }

    fun persistedByteCount(ownerIdentityId: String): Long = synchronized(fileSystemLock) {
        ownerRoot(ownerIdentityId)?.let(::thumbnailFiles)?.sumOf(File::length) ?: 0L
    }

    private fun removeStagedLocked(predicate: (SharedContentThumbnailKey) -> Boolean) {
        val iterator = stagedLookahead.entries.iterator()
        while (iterator.hasNext()) {
            val entry = iterator.next()
            if (predicate(entry.key)) {
                stagedBytes -= entry.value.size
                keyRevisions.remove(entry.key)
                iterator.remove()
            }
        }
        checkStagedAccountingLocked()
    }

    private fun checkStagedAccountingLocked() {
        check(stagedBytes >= 0)
        check(stagedBytes == stagedLookahead.values.sumOf { it.size.toLong() })
        check(stagedBytes <= MAX_STAGED_BYTES)
    }

    private fun writeDisplayedAtomically(
        key: SharedContentThumbnailKey,
        bytes: SharedContentThumbnailBytes,
    ): Boolean {
        val destination = containedFile(key) ?: return false
        if (bytes.size == 0 || bytes.size.toLong() > MAX_THUMBNAIL_BYTES) return false
        destination.parentFile?.mkdirs()
        val temporary = File(destination.parentFile, ".${destination.name}.tmp")
        return try {
            FileOutputStream(temporary, false).use { output ->
                output.write(bytes.bytes)
                output.fd.sync()
            }
            try {
                Files.move(
                    temporary.toPath(),
                    destination.toPath(),
                    StandardCopyOption.ATOMIC_MOVE,
                    StandardCopyOption.REPLACE_EXISTING,
                )
            } catch (_: AtomicMoveNotSupportedException) {
                Files.move(
                    temporary.toPath(),
                    destination.toPath(),
                    StandardCopyOption.REPLACE_EXISTING,
                )
            }
            destination.setLastModified(now().toEpochMilli())
            true
        } catch (_: Throwable) {
            false
        } finally {
            temporary.delete()
        }
    }

    private fun containedFile(key: SharedContentThumbnailKey): File? {
        val candidate = File(canonicalRoot, key.opaqueRelativePath)
        return candidate.takeIf(::isContained)
    }

    private fun ownerRoot(ownerIdentityId: String): File? =
        File(canonicalRoot, sha256(ownerIdentityId)).takeIf(::isContained)

    private fun isContained(file: File): Boolean = runCatching {
        file.canonicalFile.toPath().startsWith(canonicalRoot.toPath())
    }.getOrDefault(false)

    private fun thumbnailFiles(directory: File): List<File> =
        if (!directory.exists()) emptyList() else Files.walk(directory.toPath()).use { stream ->
            stream.iterator().asSequence().filter { path ->
                Files.isRegularFile(path) && path.fileName.toString().endsWith(".thumb")
            }.map { it.toFile() }.toList()
        }

    private fun deleteContainedTree(directory: File) {
        if (!directory.exists() || !isContained(directory)) return
        Files.walk(directory.toPath()).use { stream ->
            stream.sorted(Comparator.reverseOrder()).forEach { path ->
                val file = path.toFile()
                if (isContained(file)) file.delete()
            }
        }
    }

    private fun removeEmptyDirectories(directory: File) {
        if (!directory.exists() || !isContained(directory)) return
        Files.walk(directory.toPath()).use { stream ->
            stream.sorted(Comparator.reverseOrder()).forEach { path ->
                val file = path.toFile()
                if (file != canonicalRoot && isContained(file) && file.isDirectory && file.listFiles().isNullOrEmpty()) {
                    file.delete()
                }
            }
        }
    }

    private companion object {
        const val ROOT_DIRECTORY = "shared-content-thumbnails"
        const val MAX_THUMBNAIL_BYTES = 8L * 1024L * 1024L
        const val MAX_STAGED_BYTES = 16L * 1024L * 1024L
        const val DEFAULT_MAX_BYTES = 64L * 1024L * 1024L
        val DEFAULT_INACTIVITY: Duration = Duration.ofDays(30)

        fun sha256(value: String): String = MessageDigest.getInstance("SHA-256")
            .digest(value.toByteArray(Charsets.UTF_8))
            .joinToString("") { byte -> "%02x".format(byte.toInt() and 0xff) }
    }

    private data class PendingPromotion(
        val bytes: SharedContentThumbnailBytes,
        val keyRevision: Long,
        val ownerPurgeEpoch: Long,
        val conversationPurgeEpoch: Long,
    )
}
